### PR TITLE
Add an endpoint for resending verification emails

### PR DESCRIPTION
**Pull request by alexwlchan**
For https://github.com/wellcomecollection/wellcomecollection.org/issues/8179

I think this will allow the identity web app to make a POST to the identity API and get a new verification email to be sent. There's a separate piece of work to actually call it from the identity web app.

### CODE REVIEW

**Review by jamieparkinson — COMMENTED** [review]
This is good so far but needs the following:

- Tests
- A permission/scope to be represented in the [resource server](https://github.com/wellcomecollection/identity/blob/main/infra/scoped/auth0-resource-server.tf)
- An entry [in the authorizer](https://github.com/wellcomecollection/identity/blob/main/packages/apps/api-authorizer/src/handler.ts#L26-L47)

File: packages/apps/api/test/routes/users_user_id_resend_verification_email.test.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+describe('/users/{userId}/validate', () => {
+  describe('POST /users/{userId}/resend_verification', () => {
+    it('resends a verification email for an existing user', async () => {
+      const testUser = randomExistingUser({ password: randomAlphanumeric(10) });
+      const { api } = mockedApi([testUser]);
+      const response = await withCallerId(testUser.userId)(
+        api.post(`/users/${testUser.userId}/resend_verification`)
+      );
+
+      expect(response.statusCode).toBe(204);
```
Comment (jamieparkinson) [inline packages/apps/api/test/routes/users_user_id_resend_verification_email.test.ts:13]: See comment on mock class - you're testing the endpoint behaviour here, but not yet whether it's doing what it says it does. You can do something along the lines of...

```typescript
const { api, clients } = mockedApi([testUser]);
const response = await withCallerId(testUser.userId)(
  api.post(`/users/${testUser.userId}/resend_verification`)
);

expect(response.statusCode).toBe(204);
expect(clients.auth0.resendVerificationEmail).toHaveBeenLastCalledWith(testUser.userId);
```

File: packages/shared/auth0-client/src/MockAuth0Client.ts [hunk 2]
```diff
@@ -127,8 +127,14 @@ export default class MockAuth0Client implements Auth0Client {
...
+  private resentVerificationEmails: number[] = [];
+
   resendVerificationEmail = jest.fn(async (userId: number) => {
-    return successResponse(undefined);
+    if (this.users.has(userId.toString())) {
+      this.resentVerificationEmails.push(userId);
+      return successResponse(undefined);
+    }
+    return errorResponse('Not found', ResponseStatus.NotFound);
   });
```
Comment (jamieparkinson) [inline packages/shared/auth0-client/src/MockAuth0Client.ts:138]: What's the purpose of `resentVerificationEmails` here, as it's not being read? When you write a test for the `HttpAuth0Client` method, then you can "trust" this implementation and your test above can use the `jest.fn` to make sure it was called correctly.

Comment (alexwlchan) [reply packages/shared/auth0-client/src/MockAuth0Client.ts:138]: Ah, I think this was me trying to do what you describe in your other comment – I was planning to inspect the value and check it had been populated correctly to see that the method had been called.

Comment (alexwlchan) [reply packages/shared/auth0-client/src/MockAuth0Client.ts]: I've removed this.

File: packages/apps/api/test/routes/users_user_id_resend_verification_email.test.ts (hunk 1 above)
Comment (alexwlchan) [reply packages/apps/api/test/routes/users_user_id_resend_verification_email.test.ts:13]: This is much nicer, added.

**Comment by alexwlchan** [comment]
@jamieparkinson I think this is ready for another look.

**Review by jamieparkinson — COMMENTED** [review]
This is good, but just to be perfectionist can you [add a test](https://github.com/wellcomecollection/identity/blob/main/packages/shared/auth0-client/tests/auth0-client.test.ts) for the new method you've added to the client (I'm aware that there are not currently tests for every method, but there should be 😬)

**Comment by alexwlchan** [comment]
Test added.

I've renamed it to "send verification" email rather than "resend", because in some cases we've never sent an email in the first place.

I've also removed the scope; it turns out this is covered by the existing `update:users` scope.

**Comment by alexwlchan** [comment]
Ah, sorry, didn't realise Git was waiting for me to force push. 🤦‍♀️ 

This actually has the updated code now.

File: packages/apps/api-authorizer/src/handler.ts [hunk 3]
```diff
@@ -34,6 +34,9 @@ const validateRequest = resourceAuthorizationValidator({
   '/users/{userId}/deletion-request': {
     PUT: allOf(isSelf, hasScopes('delete:patron')),
   },
+  '/users/{userId}/send_verification': {
+    POST: isSelf,
```
Comment (jamieparkinson) [inline packages/apps/api-authorizer/src/handler.ts:38]: Why no scope check on this?

**Review by jamieparkinson — APPROVED** [review]

File: packages/apps/api-authorizer/src/handler.ts (hunk 3 above)
Comment (alexwlchan) [reply packages/apps/api-authorizer/src/handler.ts:38]: Because I forgot. 🤦‍♀️

**Pull request merged**